### PR TITLE
For Mentor Review | Concurrency | Baz Murphy

### PR DESCRIPTION
**Pull request by bazmurphy**
# Learners, PR Template

## Self checklist

- [x] I have titled my PR with `For Mentor Review | {Project Name} | {My Name}`
- [ ] I have included a list of changes **(initial commit)**
- [ ] I have tested my changes
- [ ] I have explained my changes **(initial commit)**
- [ ] I have asked someone to review my changes

## Description

This is my first attempt at `Prep` `project 05` `concurrency`

### Change list
Not applicable (initial commit)

## Rationale, briefly

- What is the goal of these changes?
Not applicable (initial commit)

- How did these changes address this goal or goals?
Not applicable (initial commit)

- Why are they a good idea?
Not applicable (initial commit)

## @mentions, #links

- Who needs to know about this? @mention them
@illicitonion Daniel

- Is this related to another issue, PR, or repo? #link them
Not applicable (initial commit)

### CODE REVIEW

**Comment by bazmurphy** [comment]
I have been bashing my head against this all day
I am opening it for review but it is not really "ready"
I don't find the generics that easy at the moment
My concurrency test is failing probably because of where I am locking/unlocking and the way I am moving the elements(nodes) around in the list
I need a bit of encouragement/fresh perspective on this

File: projects/concurrency/main.go [hunk 1]
```diff
@@ -0,0 +1,87 @@
...
+	mu       sync.Mutex
+}
+
+type CacheValue[K comparable, V any] struct {
+	// i don't like that i am duplicating the key here... but it is for easier lookup later
+	key   K
+	value V
+}
+
+func NewCache[K comparable, V any](entryLimit int) Cache[K, V] {
```
Comment (illicitonion) [inline projects/concurrency/main.go:21]: What happens if you pass 0 or -1 as the `entryLimit` and then try to use the cache?

File: projects/concurrency/main_test.go [hunk 2]
```diff
@@ -0,0 +1,184 @@
...
+		insertKey := "a"
+		insertValue := 100
+
+		ok := cache.Put(insertKey, insertValue)
+
+		if ok != false {
+			t.Errorf("found an existing value on the key in the cache, where there should have been no existing value")
+		}
+
+		cacheListElementPointer, ok := cache.entries[insertKey]
```
Comment (illicitonion) [inline projects/concurrency/main_test.go:32]: In general we prefer to test against the public API of our types where possible (i.e. calling `cache.Get` rather than looking in `cache.entries`) unless there's a specific internal property of the type we're looking to assert on.

This has a few advantages:
* If we change the internals of the type (e.g. switching from a linked list to some other data structure, or changing the internals of `CacheValue`) the test will still be valid. It's generally _really_ nice if we can make internal changes to an implementation detail without needing to change existing tests, it helps us have confidence we haven't changed the behaviours the tests were there to test.
* Serves as an example of how the type is intended to be used
* Often easier to verify that the tests are correctly written, because the test is all written at one layer of abstraction rather than crossing an abstraction layer.

File: projects/concurrency/main_test.go [hunk 3]
```diff
@@ -0,0 +1,184 @@
...
+			t.Errorf("could not find key %v when there should be one", insertKey)
+		}
+
+		// (!) deference the pointer to a value
+		if *valuePointer != updatedValue {
+			t.Errorf("got %v | want %v", *valuePointer, updatedValue)
+		}
+	})
+
+	t.Run("add 3 keys to an empty cache", func(t *testing.T) {
```
Comment (illicitonion) [inline projects/concurrency/main_test.go:80]: I'd probably call the test "fill an empty cache" rather than hard-coding 3 - 3 isn't really a magic number in terms of the _interface_ of the cache that you're testing, it's an implementation detail about what "full" means for the cache you happen to be making.

File: projects/concurrency/main_test.go [hunk 4]
```diff
@@ -0,0 +1,184 @@
...
+		cache := NewCache[string, int](3)
+
+		insertKeyValues := map[string]int{"a": 1, "b": 2, "c": 3}
+
+		for insertKey, insertValue := range insertKeyValues {
```
Comment (illicitonion) [inline projects/concurrency/main_test.go:85]: Iterating a `map` in Go has undefined ordering, and will actually return in different orders in different attempts to iterate. If you're making assertions about ordering, you probably don't want to be using a `map` for your operations.

File: projects/concurrency/main_test.go [hunk 5]
```diff
@@ -0,0 +1,184 @@
...
+			cacheValue := cacheListElementPointer.Value.(*CacheValue[string, int])
+
+			if cacheValue.value != insertValue {
+				t.Errorf("cacheValue.value: got %v want %v", cacheValue.value, insertValue)
+			}
+		}
+
+		expectedCacheListOrder := []int{3, 2, 1}
+
+		cacheListElementPointer := cache.list.Front()
```
Comment (illicitonion) [inline projects/concurrency/main_test.go:106]: Can you think how you could test eviction order _without_ looking in the internals of the cache, just using the public interface?

File: projects/concurrency/main_test.go [hunk 6]
```diff
@@ -0,0 +1,184 @@
...
+
+			cacheListElementPointer = cacheListElementPointer.Next()
+		}
+	})
+}
+
+// note: need to add another test here for 6 keys to an empty cache, to check the LRU order
+
+// this test is failing, probably because of the way i am concurrently accessing the cache
+func TestCacheConcurrency(t *testing.T) {
```
Comment (illicitonion) [inline projects/concurrency/main_test.go:124]: What concurrency-safe property of your cache is this test attempting to assert? As far as I can tell, it's testing that:

Given a cache that can hold 3 entries
If 10 goroutines each try to write different values to keys that only that goroutine writes to, they will a short amount of time later be able to read back the value that was written

I don't think this is a property promised by our cache interface. I can imagine a cache which does offer such a property, but it would probably require a concept of a transaction, i.e. you'd say to the cache "I'm about to start giving you a sequence of operations and I want a consistent view of the cache until I tell you I'm done", some code like below:

```go
txn := cache.BeginTransaction()
txn.Put("key-0", 0)
value, ok := txn.Get("key-0")
require.True(t, ok)
require.Equal(t, value, 0)
txn.End()
```

Give some careful thought to what promises we _do_ expect to hold. I think the property you're testing here holds _if the value is present_ (i.e. concurrent writes to _different_ keys shouldn't change independent keys), but it's also valid for the key being looked up to be missing (because three other goroutines did writes between this goroutine's write and read). How could you modify your test to test this property instead?

Another example of a promise I think we do expect is: If a bunch of goroutines do a bunch of writes to different keys, then afterwards exactly three keys should be present in the cache, with the last values written to those keys. One of the tricky things with testing concurrent behaviour is that we don't know _which_ three keys. So we need to write a less restrictive test - we do know some properties of those keys though - we know they should be unique (because our cache promises it can hold three unique keys), we know which keys _could be_ expected (in your example set-up here we know we shouldn't see "key-15", for instance), we know they should have the last values written to them (so in your example set-up here we know the value for "key-0" should be 10, if it's present).

There are also tests we can write e.g. where multiple goroutines are writing different values to the same key - again, we can't write _precise_ assertions on what values we expect to be in the cache, but we can assert on properties of what we do find in the cache.

(Also, if you find yourself interested, try writing the transaction-style cache described above! It's a really interesting problem actually! One of the interesting trade-offs is: Where do you want to block? If someone tries to overwrite a value that some other transaction needs to be able to retrieve, do you want to block the write until the other transaction completes? Do you want to keep two different values for the key which are visible to different transactions - if so do you want to do so inside the cache itself, or do you want each transaction to have a list of "pending" writes? Do you want to reject the write completely? Do you want to allow reads/writes from outside transactions, or only inside them? _All_ of those answers may be valid, and others too, but each suggests a different design of the internals. What examples of use-cases might each of those behaviours be useful for? Some example code which illustrates the problem is below)

```go

cache := NewCache()

var wg sync.WaitGroup

wg.Add(2)

timeBefore := time.Now()

go func() {
    txn := cache.BeginTxn()
    txn.Put("key-0", 0)
    time.Sleep(5 * time.Second)
    value := txn.Get("key-0")
    fmt.Printf("From goroutine 1 expect value 0, got: %d\n", value)
    txn.End()
    wg.Done()
}()

go func() {
    txn := cache.BeginTxn()
    txn.Put("key-0", 1)
    time.Sleep(5 * time.Second)
    value := txn.Get("key-0")
    fmt.Printf("From goroutine 1 expect value 1, got: %d\n", value)
    txn.End()
    wg.Done()
}

wg.Wait()

fmt.Printf("Took %v to wait for two goroutines, value of key-0 was %v\n", time.Now().Sub(timeBefore), cache.Get("key-0"))
```

File: projects/concurrency/main.go [hunk 7]
```diff
@@ -0,0 +1,87 @@
...
+import (
+	"container/list"
+	"sync"
+)
+
+type Cache[K comparable, V any] struct {
+	capacity int
+	entries  map[K]*list.Element
+	list     *list.List
+	mu       sync.Mutex
```
Comment (illicitonion) [inline projects/concurrency/main.go:12]: I'd generally recommend stating in comments which fields are guarded by this mutex. I think in your example here it's `entries` and `list`?

It's also fairly common to show this just by the order and groupings of fields in a struct, so something like:

```go
type Cache[K comparable, V any] struct {
	capacity int

	mu sync.Mutex
	entries  map[K]*list.Element
	list     *list.List
}
```

File: projects/concurrency/main.go [hunk 8]
```diff
@@ -0,0 +1,87 @@
...
+}
+
+type CacheValue[K comparable, V any] struct {
+	// i don't like that i am duplicating the key here... but it is for easier lookup later
```
Comment (illicitonion) [inline projects/concurrency/main.go:16]: This seems fine. If you're worried about potentially using a lot of memory, you could keep a pointer to the key here, so you're not making a copy.

**Review by illicitonion — COMMENTED** [review]
Your implementation generally looks solid here - I have mostly left some questions about trade-offs and style, rather than correctness.

The single-threaded tests look fine, but have left some thoughts on the concurrency test.

I'd recommend adding the stats counters piece of the project, and fleshing out some of the concurrency tests, then either trying the transaction-aware cache (which will require you to think about how you structure the program a bit), or the computing-cache described in the exercise (which will require you to think about how to avoid blocking unrelated threads) - both are pretty interesting!
